core/libp2p/protocol_muxer/multiselect/multiselect.cpp

```diff
@@ -0,0 +1,283 @@
+                           kagome::common::Logger logger)
+      : peer_id_{std::move(peer_id)}, log_{std::move(logger)} {}
+
+  Multiselect::~Multiselect() = default;
+
+  void Multiselect::addProtocol(const Protocol &protocol) {
+    supported_protocols_.push_back(protocol);
+  }
+
+  void Multiselect::negotiateServer(const stream::Stream &stream,
+                                    ChosenProtocolCallback protocol_callback) {}
+
+  void Multiselect::negotiateClient(const stream::Stream &stream,
+                                    ChosenProtocolCallback protocol_callback) {
+    // if no protocols are supported by this instance, there's no sense in
+    // starting a negotiation process
+    if (supported_protocols_.empty()) {
+      protocol_callback(MultiselectErrors::NO_PROTOCOLS_SUPPORTED);
+      return;
+    }
+    sendOpeningMsg(StreamState{std::cref(stream), std::move(protocol_callback),
+                               StreamState::NegotiationStatus::OPENING_SENT});
+  }
+
+  void Multiselect::readResponse(StreamState stream_state) const {
+    stream_state.stream_.get().readAsync(
+        [t = shared_from_this(), stream_state = std::move(stream_state)](
+            stream::Stream::NetworkMessageOutcome response) mutable {
+          if (!response) {
+            t->log_->error("cannot read a response from the stream: {}",
+                           response.error().value());
+            stream_state.proto_callback_(response.error());
+            return;
+          }
+          t->processResponse(response.value(), std::move(stream_state));
+        });
+  }
+
+  void Multiselect::processResponse(const kagome::common::Buffer &response,
+                                    StreamState stream_state) const {
+    using MessageType = MessageManager::MultiselectMessage::MessageType;
+
+    auto message_res = MessageManager::parseMessage(response);
+    if (!message_res) {
+      log_->error("cannot parse message, received from the stream: {}",
+                  message_res.error().value());
+      // response by sending ls; this can still allow us to continue negotiation
+      sendLsMsg(std::move(stream_state));
+      return;
+    }
+
+    auto message = std::move(message_res.value());
+
+    // memorize peer_id of the other side, if we don't know it yet
+    stream_state.setPeerId(std::move(message.peer_id_));
+
+    switch (message.type_) {
+      case MessageType::OPENING:
+        handleOpeningMsg(std::move(stream_state));
+        return;
+      case MessageType::PROTOCOL:
+        handleProtocolMsg(message.protocols_[0], std::move(stream_state));
+        return;
+      case MessageType::PROTOCOLS:
+        handleProtocolsMsg(message.protocols_, std::move(stream_state));
+        return;
+      case MessageType::LS:
+        handleLsMsg(std::move(stream_state));
+        return;
+      case MessageType::NA:
+        handleNaMsg(std::move(stream_state));
+        return;
+    }
```

Comment:
you should handle also default case, as message may contain garbage

Reply:
Garbage, I think, would never be returned by our parse function. But OK, I will add the handler